src/ray/raylet/node_manager.cc

```diff
@@ -835,14 +838,20 @@ void NodeManager::ProcessClientMessage(
 void NodeManager::ProcessRegisterClientRequestMessage(
     const std::shared_ptr<LocalClientConnection> &client, const uint8_t *message_data) {
   auto message = flatbuffers::GetRoot<protocol::RegisterClientRequest>(message_data);
-  client->SetClientID(from_flatbuf<ClientID>(*message->worker_id()));
+  auto client_id = from_flatbuf<ClientID>(*message->worker_id());
+  client->SetClientID(client_id);
   Language language = static_cast<Language>(message->language());
-  auto worker =
-      std::make_shared<Worker>(message->worker_pid(), language, message->port(), client);
+  auto worker = std::make_shared<Worker>(message->worker_pid(), language, message->port(),
+                                         client, client_call_manager_);
   if (message->is_worker()) {
     // Register the new worker.
+    bool use_push_task = worker->UsePush();
+    auto connection = worker->Connection();
     worker_pool_.RegisterWorker(std::move(worker));
-    DispatchTasks(local_queues_.GetReadyTasksWithResources());
+    if (use_push_task) {
+      // only call `ProcessGetTaskMessage` when push mode is used.
+      ProcessGetTaskMessage(connection);
```

Comment:
Nit, `ProcessGetTaskMessage`'s name is kind of improper here. 
I would rename it to something like `OnWorkerIdle` and pass the worker as the parameter.

Reply:
I agree that the naming here is not best, but also I'm a bit hesitant to add another function which is exactly the same as `ProcessGetTaskMessage` with just a different name. One the other hand, we already call `ProcessDisconnectClientMessage` when we want to disconnect a worker, so it might also be ok here.

Reply:
I think we can just rename `ProcessGetTaskMessage`. To match the rest of the code, something like `HandleWorkerAvailable` would be good (I think similar to what you had in a previous commit).